source/MaterialXRuntime/RtCommandEngine.cpp

```diff
@@ -0,0 +1,118 @@
+//
+// TM & (c) 2020 Lucasfilm Entertainment Company Ltd. and Lucasfilm Ltd.
+// All rights reserved.  See LICENSE.txt for license.
+//
+
+#include <MaterialXRuntime/RtCommandEngine.h>
+#include <MaterialXRuntime/Private/PvtCommandEngine.h>
+
+namespace MaterialX
+{
+
+namespace
+{
+    // Syntactic sugar
+    using RtCommandPtrVec = vector<RtCommandPtr>;
+    inline RtCommandPtrVec* _getvec(void* ptr)
+    {
+        return static_cast<RtCommandPtrVec*>(ptr);
+    }
+}
+
+RtBatchCommand::RtBatchCommand() :
+    _ptr(new vector<RtCommandPtr>)
+{
+}
+
+RtBatchCommand::~RtBatchCommand()
+{
+    delete _getvec(_ptr);
+}
+
+void RtBatchCommand::addCommand(RtCommandPtr cmd)
+{
+    _getvec(_ptr)->push_back(cmd);
+}
+
+void RtBatchCommand::clearCommands()
+{
+    _getvec(_ptr)->clear();
+}
+
+void RtBatchCommand::execute(RtCommandResult& result)
+{
+    for (RtCommandPtr cmd : *_getvec(_ptr))
+    {
+        cmd->execute(result);
+        if (!result.success())
+        {
+            break;
+        }
+    }
+}
+
+void RtBatchCommand::undo(RtCommandResult& result)
+{
+    for (RtCommandPtr cmd : *_getvec(_ptr))
+    {
+        cmd->undo(result);
+        if (!result.success())
+        {
+            break;
```

Comment:
What does it mean to partially undo / or redo something? Isn't it possible to leave things in a bad state afterwards?

Reply:
This batch of command is manly meant to be used when several commands is needed to perform an operation (for example breaking connectinos before removing a prim). One one command in that list fails I don't think we should continuing with the rest, since they are depending on each other. So I think it's better to stop and report the error.